vhdx/src/vhdx_metadata.rs

```diff
@@ -0,0 +1,323 @@
+// Copyright © 2021 Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0 AND BSD-3-Clause
+
+use crate::vhdx_header::RegionTableEntry;
+use byteorder::{ByteOrder, LittleEndian, ReadBytesExt};
+use remain::sorted;
+use std::convert::TryInto;
+use std::fmt::{self, Display};
+use std::fs::File;
+use std::io::{self, Read, Seek, SeekFrom};
+use std::mem::size_of;
+
+const METADATA_SIGN: u64 = 0x6174_6164_6174_656D;
+const METADATA_ENTRY_SIZE: usize = 32;
+const METADATA_MAX_ENTRIES: u16 = 2047;
+// The size including the table header and entries
+const METADATA_TABLE_MAX_SIZE: usize = METADATA_ENTRY_SIZE * (METADATA_MAX_ENTRIES as usize + 1);
+
+const METADATA_FLAGS_IS_REQUIRED: u32 = 0x04;
+
+pub const BLOCK_SIZE_MIN: u32 = 1024 * 1024; // 1 MiB
+const BLOCK_SIZE_MAX: u32 = 256 * 1024 * 1024; // 256 MiB
+const MAX_SECTORS_PER_BLOCK: u64 = 1 << 23;
+
+const BLOCK_HAS_PARENT: u32 = 0x02; // has a parent or backing file
+
+// GUIDs when read in Little Endian u128 integer
+const METADATA_FILE_PARAMETER: u128 = 143428208870287139673581922146496636727; //"CAA16737-FA36-4D43-B3B6-33F0AA44E76B";
+const METADATA_VIRTUAL_DISK_SIZE: u128 = 245846085492271402621565837099086987812; //"2FA54224-CD1B-4876-B211-5DBED83BF4B8";
+const METADATA_VIRTUAL_DISK_ID: u128 = 94079244529923181982015449247999726251; //"BECA12AB-B2E6-4523-93EF-C309E000C746";
+const METADATA_LOGICAL_SECTOR_SIZE: u128 = 127169626291185313037823917095022346013; //"8141BF1D-A96F-4709-BA47-F233A8FAAB5F";
+const METADATA_PHYSICAL_SECTOR_SIZE: u128 = 115338139532893941197488005837198936263; //"CDA348C7-445D-4471-9CC9-E9885251C556";
```

Comment:
Ditto about uuid crate.

Reply:
There are two problems. 
1. According to the spec, we are supposed to read a UUID as 128bit integer in little-endian format. However, this uuid crate converts UUIDs to 128 bit integers in big-endian format, which gives different values. Therefore, I manually figured the corresponding little-endian values for the UUIDs and hard coded here.

2.  We have a match statement where check if the parsed integer value matches one of the known UUIDs. Match statement only takes constants for patterns. If we had used uuid crate to convert the UUIDs to integer, then we would get variables, which are not acceptable in match. Therefore, we realized it is easier and simpler to hard code the UUID values in integers while formatted in little-endian.

Reply:
We using uuid crate now to convert from UUID to u128.